src/globals.rs

```diff
@@ -0,0 +1,47 @@
+use crate::error::GlobalError;
+use wayland_client::Proxy;
+
+/// A trait implemented by types that provide access to capability globals.
+///
+/// The returned global must be fully compatible with the provided `API_COMPAT_VERSION` generic
+/// argument.  For example:
+///
+/// - A global that binds to `wl_compositor` with maximum version 4 could implement
+/// `ProvidesBoundGlobal<WlCompositor, 4>`, `ProvidesBoundGlobal<WlCompositor, 3>`,
+/// `ProvidesBoundGlobal<WlCompositor, 2>`, and `ProvidesBoundGlobal<WlCompositor, 1>` because
+/// versions 2-4 only add additional requests to the `wl_surface` API.
+/// - A global that binds to `wl_compositor` with maximum version 5 may only implement
+/// `ProvidesBoundGlobal<WlCompositor, 5>` because version 5 makes using `wl_surface::attach` with
+/// a nonzero offset a protocol error.  A caller who is only aware of the version 4 API risks
+/// causing these protocol errors if it uses surfaces created by such a global.
```

Comment:
Cc @vberger

Reply:
Sounds like a reasonable design principle to me, only provide multiple impl if there is no forward-compatibility hazard for the downstream user.

Reply:
Hmm, I was under the impression on that doc test that it is ok to implement `ProvidesBoundGlobal` on `WlCompositor` from version 1 to 4 because essentially, if downstream requests version 1 and you provide version 4 instead, nothing will break in any way.

If providing `ProvidesBoundGlobal<WlCompositor, 2>` means you promise to not provide a high version than 2 even if they would be compatible, that means you cannot do early binding of the global, because you don't know which version the user will request.

Reply:
That's mostly correct. Providing `ProvidesBoundGlobal<WlCompositor, 2>` means that one of the following is true:

- You are providing a global whose maximum version is `<= 2`.  This could happen because the compositor only supports version 1, or it could be because you only bind version 1 even if version 2 (or 4) is available.
- You are providing a global which is API-compatible with version `2`. This requires that you look at the changes and make sure nothing will break.

This definition permits current code to implement `ProvidesBoundGlobal<WlCompositor, 6>` even if version 6 has not yet been defined, because code that is capable of supporting version 6 will be able to use the global that the current code binds - they might be versions 1-5, all of which someone requesting version 6 needs to be prepared to handle in case the compositor itself only supports that version.  This means that you can release a crate that uses newer features without waiting on an upgrade of both SCTK and wayland-protocols, but which will be able to use them once SCTK has updated.